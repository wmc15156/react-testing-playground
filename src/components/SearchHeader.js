import React, { useEffect, useState } from "react";
import { useNavigate, useParams } from "react-router-dom";

export default function SearchHeader() {
  const { keyword } = useParams();
  const [search, setSearch] = useState("");
  const navigate = useNavigate();

  const handleInputChange = (event) => {
    setSearch(event.target.value);
  };

  useEffect(() => {
    setSearch(keyword || "");
  }, [keyword]);

  const handleSearch = () => {
    navigate(`/products/${search}`);
  };
  console.log(12123);
  return (
    <div>
      <input type="text" value={search} onChange={handleInputChange} />
      <button onClick={handleSearch}>Search</button>
    </div>
  );
}
