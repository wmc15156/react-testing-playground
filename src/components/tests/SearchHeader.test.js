import renderer from "react-test-renderer";
import SearchHeader from "../SearchHeader";
import { withMemoryRouter } from "./utils";
import { Route } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";

describe("Search Header", () => {
  it("renders snapshot", () => {
    const searchHeader = renderer
      .create(withMemoryRouter(<Route path="/" element={<SearchHeader />} />))
      .toJSON();

    expect(searchHeader).toMatchSnapshot();
  });

  it("url로 접속했을때 정상적으로 입력창에 keyword가 들어가는지 확인", () => {
    // given
    render(
      withMemoryRouter(
        <Route path="/:keyword" element={<SearchHeader />} />,
        "/shoes"
      )
    );

    expect(screen.getByDisplayValue("shoes")).toBeInTheDocument();
  });

  it("인풋 버튼을 클릭했을때 products 페이지로 정상적으로 이동", () => {
    const fakeSearch = "shoes";

    render(
      withMemoryRouter(
        <>
          <Route path="/" element={<SearchHeader />} />,
          <Route
            path={`/products/:keyword`}
            element={<span data-testid="search-result" />}
          />
        </>
      )
    );

    const searchBtn = screen.getByRole("button");
    const searchInput = screen.getByRole("textbox");

    fireEvent.change(searchInput, { target: { value: fakeSearch } });
    fireEvent.click(searchBtn);

    expect(screen.getByTestId("search-result")).toBeInTheDocument();
  });
});
