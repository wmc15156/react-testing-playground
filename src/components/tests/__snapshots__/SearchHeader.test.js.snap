// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Header renders snapshot 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Search
  </button>
</div>
`;
